ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java

```diff
@@ -208,7 +201,38 @@ private void addAttributesFromPayload(Map<String, String> attributes, ObjectNode
           .ifPresent(v -> attributes.put(Attribute.OS_VERSION, v));
     }
 
+    ParsePayload.addClientIdFromPayload(attributes, json);
+
+    // Add sample id, usually based on hashing clientId, but some other IDs are also supported to
+    // allow sampling on non-telemetry pings.
+    Stream.of(attributes.get(Attribute.CLIENT_ID),
+        // "impression_id" is a client_id-like identifier used in activity-stream ping
+        // that do not contain a client_id.
+        json.path("impression_id").textValue()) //
+        .map(ParsePayload::normalizeUuid) //
+        .filter(Objects::nonNull) //
+        .findFirst() //
+        .ifPresent(v -> attributes.put(Attribute.SAMPLE_ID, Long.toString(calculateSampleId(v))));
+  }
+
+  /** Extracts the client ID from the payload and adds it to `attributes`. */
+  public static void addClientIdFromPayload(Map<String, String> attributes, ObjectNode json) {
```

Comment:
I think this should be refactored to return an `Optional<String>` rather than setting it in `attributes`.
```suggestion
  public static Optional<String> getNormalizedClientId(ObjectNode json) {
```
that way `bug1489560Affected` doesn't pull it out of a temporary hashmap or try to modify `gleanClientInfo`.

in `addAttributesFromPayload` the call to `getNormalizedClientId` would need to occur before `if (gleanClientInfo.isObject()) {`, and be used to set client_id within `gleanClientInfo` in the same place as before this PR.

Reply:
The problem here is that not only the glean ping needs to get potentially updated, but also the common ping or pings sending `clientId`. It's possible to return the client ID, but I don't think modifying the json structure can be avoided without duplicating the logic. I felt like it felt more intuitive to modify the attributes since it's more similar to `addAttributesFromPayload`.

Reply:
I agree with @scholtzan that having this as a side-effect is the practical thing to do here. I think we can rely on copious documentation to explain that we're doing something weird here and why.